cmd/lotus-shed/chain_index.go

```diff
@@ -0,0 +1,222 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/ipfs/go-cid"
+	"github.com/urfave/cli/v2"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+	lcli "github.com/filecoin-project/lotus/cli"
+)
+
+func withCategory(cat string, cmd *cli.Command) *cli.Command {
+	cmd.Category = strings.ToUpper(cat)
+	return cmd
+}
+
+var chainIndexCmds = &cli.Command{
+	Name:            "chainindex",
+	Usage:           "Commands related to managing the chainindex",
+	HideHelpCommand: true,
+	Subcommands: []*cli.Command{
+		withCategory("chainindex", validateBackfillChainIndexCmd),
+	},
+}
+
+var validateBackfillChainIndexCmd = &cli.Command{
+	Name:  "validate-backfill",
+	Usage: "Validates and optionally backfills the chainindex for a range of epochs",
+	Description: `
+lotus-shed chainindex validate-backfill --from <start_epoch> --to <end_epoch> [--backfill] [--log-good]
+
+The command validates the chain index entries for each epoch in the specified range, checking for missing or
+inconsistent entries (i.e. the indexed data does not match the actual chain state). If '--backfill' is enabled
+(which it is by default), it will attempt to backfill any missing entries using the 'ChainValidateIndex' API.
+
+Parameters:
+  - '--from' (required): The starting epoch (inclusive) for the validation range. Must be greater than 0.
+  - '--to' (required): The ending epoch (inclusive) for the validation range. Must be greater than 0 and less
```

Comment:
From chatting with @jennijuju just now - we want to test this _before_ FEVM height because a full-archival node will probably want backfill to genesis to get msglookup, I think our FEVM archival node has history before that so we should test those epochs before Hygge that we have in there.

Reply:
In terms of events, I believe that it should be transparent to the indexer, https://github.com/filecoin-project/lotus/blob/c96f4edd0a5388887e1b24834682c4f7ab3f1484/chain/types/message_receipt_cbor.go#L60-L67 receipts have a version break at FIP-0049 where they add the new `EventRoot` CID field and receipts older than FIP-0049 don't have this encoded but they instantiate with that field being empty, and we have checks where it matters for `EventRoot == nil`.

But it would be good to test some pre-FIP-0049 tipsets.

Reply:
Okay will tests this when we re-do archival node testing  once all the reviews on this one have been addressed.

Reply:
```
ubuntu@ip-192-168-16-104:~/lotus$ ./lotus-shed chainindex validate-backfill --from 2683347 --to 2683347 --log-good
2024-10-21 08:27:42.864 starting chainindex validation; from epoch: 2683347; to epoch: 2683347; backfill: true; log-good: true
2024-10-21 08:27:42.864 -------- Chain index validation progress: 100.00%; Time elapsed: 33.648µs
2024-10-21 08:27:45.308 ✓ Epoch 2683347 ({"TipSetKey":[{"/":"bafy2bzacebziho7glfpn54gusepa6d6akzagn7rthhgd5bqk2enxwq5gzg3gy"},{"/":"bafy2bzacec5eg6ik25nvkxf2gvlspgiqcvzqkhapvaaqkt5uzsdlkuctgygsy"},{"/":"bafy2bzaceawlnbnbszpz5u2ji7fbxy4z2rhuk6mxjo3dddlvaejoi3wjsvmok"}],"Height":2683347,"IndexedMessagesCount":174,"IndexedEventsCount":0,"IndexedEventEntriesCount":0,"Backfilled":true,"IsNullRound":false})

2024-10-21 08:27:45.308 Chain index validation summary:
Total failed RPC calls: 0
Total successful backfills: 1
Total successful validations without backfilling: 0
Total successful Null round validations: 0
```

Reply:
`eth_getBlockReceipts` RPC also returns a valid response with empty logs for all.

Reply:
@rvagg ^^ Tested backfilling for an epoch before Hygge launch and it works.